galley/pkg/runtime/projections/serviceentry/handler.go

```diff
@@ -0,0 +1,311 @@
+	pendingEvents int64
+
+	// lastSnapshotTime records the last time a snapshot was published.
+	lastSnapshotTime time.Time
+}
+
+// NewHandler creates a new Handler instance.
+func NewHandler(domainSuffix string, listener processing.Listener) *Handler {
+	pods, podsHandler := pod.NewCache()
+	nodes, nodesHandler := node.NewCache()
+
+	return &Handler{
+		domainSuffix: domainSuffix,
+		listener:     listener,
+		services:     make(map[resource.FullName]resource.Entry),
+		endpoints:    make(map[resource.FullName]resource.Entry),
+		mcpResources: make(map[resource.FullName]*mcp.Resource),
+		podsHandler:  podsHandler,
+		nodesHandler: nodesHandler,
+		pods:         pods,
+		nodes:        nodes,
+	}
+}
+
+// Handle incoming events and generate synthetic ServiceEntry projections.
+func (p *Handler) Handle(event resource.Event) {
+	switch event.Entry.ID.Collection {
+	case metadata.K8sCoreV1Endpoints.Collection:
+		// Update the projections
+		p.handleEndpointsEvent(event)
+	case metadata.K8sCoreV1Services.Collection:
+		// Update the projections
+		p.handleServiceEvent(event)
+	case metadata.K8sCoreV1Nodes.Collection:
+		// Just add the node to the cache.
+		p.nodesHandler.Handle(event)
+	case metadata.K8sCoreV1Pods.Collection:
+		// Just add the pod to the cache.
+		p.podsHandler.Handle(event)
+	default:
+		scope.Warnf("received event with unexpected collection: %v", event.Entry.ID.Collection)
+	}
+}
+
+// Builds the snapshot of the current resources.
+func (p *Handler) BuildSnapshot() snapshot.Snapshot {
+	p.mcpLock.Lock()
+	defer p.mcpLock.Unlock()
+
+	now := time.Now()
+	monitoring.RecordProcessorSnapshotPublished(p.pendingEvents, now.Sub(p.lastSnapshotTime))
+	p.lastSnapshotTime = now
+	p.pendingEvents = 0
+
+	b := snapshot.NewInMemoryBuilder()
+
+	// Copy the entries.
+	entries := make([]*mcp.Resource, 0, len(p.mcpResources))
+	for _, r := range p.mcpResources {
+		entries = append(entries, r)
+	}
+
+	// Create the collection resources on the Snapshot.
+	version := fmt.Sprintf("%d", p.version)
+	b.Set(collection.String(), version, entries)
+
+	return b.Build()
+}
+
+func (p *Handler) handleServiceEvent(event resource.Event) {
+	entry := event.Entry
+	fullName := entry.ID.FullName
+
+	switch event.Kind {
+	case resource.Added, resource.Updated:
+
+		// Check to see if the service's version has changed.
+		curEntry, ok := p.services[fullName]
+		if ok && curEntry.ID.Version == entry.ID.Version {
+			scope.Debugf("received service for the current, known version: %v", event)
+			return
+		}
+
+		// Store the service.
+		p.services[fullName] = entry
+
+		// Get the associated endpoints, if available.
+		var endpoints *coreV1.Endpoints
+		if endpointsEntry, ok := p.endpoints[fullName]; ok {
+			endpoints = endpointsEntry.Item.(*coreV1.Endpoints)
+		}
+
+		// Convert to an MCP resource to be used in the snapshot.
+		mcpEntry, ok := p.toMcpResource(entry, endpoints)
+		if !ok {
+			return
```

Comment:
Please log an error here.

Reply:
the `toMcpResource` method already does the logging for us.